Cli sorting options

NickHackman opened the pull request:
The use of `src/sort.rs/Sorter` structure should be fully suitable for future additions and sorting new globs during execution. It has been added to `src/program/mod.rs/Program` structure for future use.

Additions:
Command line sorting options detailed in #41, are case insensitive
Created `src/sort.rs` to have all sort related functionality namely `Sorter` and `SortOrder`
Reverse sorting CLI flag via `-r` and `--reverse`
Max number of images CLI flag via `-m` and `--max`
Dependency [natord](https://github.com/lifthrasiir/rust-natord) for natural ordering of alphanumeric filenames
Documentation to `README.md` to provide explanation of sorting options and other added command line flags

Specifics:
In commit `d34a229`, allowed missing docs is used due to the macro `arg_enum!` from clap not allowing documentation for variants inside the macro call further discussion [here](https://github.com/clap-rs/clap/issues/1473) only used in `src/sort.rs`. This was decided over manually implementing `FromStr` for the enum `SortOrder`.

Please be gentle

NickHackman committed:
Add sorting commandline options to Clap

* Reverse command line flag
* Added macro_use to clap
* Allowed missing docs in cli.rs as a result of Clap arg_enum!
    bug/restriction on enum variants

NickHackman committed:
Add Sorting for files, added dependency natord

* Dependency Natord added in order to properly sort files by name in the
    same fashion ls/exa does
* Added flag max, dictates the maximum number of images
* Sorts Breadth, Depth, Date, Alphabetical, Size
* Manipulation of Sorter at runtime

Note: Breadth and Depth each comparison requires a linear walk up their
parent directories, only fix would be to cache each path's depth.

NickHackman committed:
Add Readme docs for sort, reverse, and max_length

* Added examples
* Table for sorting with descriptions

NickHackman commented:
Very sorry, but it looks like the build failed because I made a formatting error during merging. Last time I use github to merge, need RustFmt. :frowning_face:

gurgalex commented on README.md:
````diff
@@ -27,6 +27,28 @@ Set a destination folder for moving files with the `f` flag. The folder will be
+
+```$ riv -s alphabetical **/*.png```
+
+### Sorting Options
````
Good idea to explain the sorting options.

I do wonder if Clap and/or a man page generator has an option to express these details without looking at the project readme..

gurgalex commented on src/sort.rs:
```diff
@@ -0,0 +1,131 @@
+        }
+        if self.max_length > 0 {
+            let new_length = pathes.len().saturating_sub(pathes.len() - self.max_length);
+            pathes.truncate(new_length);
```
An issue that may need addressed later is needing to rescan all the paths if the user wants to reverse the order at runtime.

Addressing that would be more complicated in needing to keep around all the other images in `images`, but capping every other option which concerns amount of images at `max_length` amount of them.

NickHackman replied:
> An issue that may need addressed later is needing to rescan all the paths if the user wants to reverse the order at runtime.
> 
True, it would be more versatile to separate these, or at least include reverse as it's own function.

In that case would it be good to move `max_length` from `Sorter` to `Path` and calculate the user viewed length upon changes of `max_length`, or the image path as a whole? That way we have an artificial cap, the user just won't see the images past `max_length` and that allows us to sort them later on.

Inexperience question:
If I were to commit these on my branch would I have to open up a new pull request or would the pull request be updated?

gurgalex replied:
> True, it would be more versatile to separate these, or at least include reverse as it's own function.

I thought of that.  The below code seems to not require adding your own, but reusing `images` reverse().

```rust
    pub fn sort<'a>(&self, pathes: &'a mut Vec<PathBuf>) -> &'a mut Vec<PathBuf> {
        pathes.sort_by(|a, b| self.sort_order.file_compare(&a, &b));
        if self.reverse {
            pathes.reverse();
        }
        if self.max_length > 0 {
            let new_length = pathes.len().saturating_sub(pathes.len() - self.max_length);
            pathes.truncate(new_length);
        }
        pathes
    }
```

So then you can do.

```rust
        sorter.sort(&mut images).reverse().truncate(22);
```


I'd definitely go with the optional `max_length` on Paths.  A different name such as `max_viewable` perhaps.

> If I were to commit these on my branch would I have to open up a new pull request or would the pull request be updated?

No. Pushing to the branch of this pull request will update this pull request.

NickHackman replied:
That looks like a great ease of use improvement onto sort, implementing that and the paths update now.

gurgalex reviewed (approved):
Looks good to me.

@Davejkane what do you think?

NickHackman committed:
Refractored Sorter, moved max_length to paths

* Paths now has max_viewable which is an artificial limit on the number
    of visible images for the user. max_viewable should be treated as
    the new path.images.len()
* Sorter::sort now returns the vector it mutates for better usability

NickHackman committed:
Merge branch 'development' into cli-sorting-options

NickHackman commented:
Ran it through RustFmt that time, man that tool is a amazing :smiley:

gurgalex commented on src/sort.rs:
```diff
@@ -0,0 +1,128 @@
+        self.reverse = reverse;
+    }
+
+    /// Sorts the images based on sort_order, reverses if necessary, and truncates as well
```
Small doc fix.
Truncating is no longer done by the sorter.

NickHackman replied:
Fixed

gurgalex commented on src/sort.rs:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    /// Reverses provided vector, Added for extra versatility
+    pub fn reverse(&self, pathes: &mut Vec<PathBuf>) {
```
Will this be more versatile in the future?

This won't reverse the vector if the user set `reverse` to false.

NickHackman replied:
Yeah, true it's unnecessary

NickHackman committed:
Fixed docs in sort.rs, removed Sorter::reverse

gurgalex commented on src/program/mod.rs:
```diff
@@ -44,8 +46,21 @@ impl<'a> Program<'a> {
         args: cli::Args,
     ) -> Result<Program<'a>, String> {
-        let images = args.files;
+        let args = cli::cli()?;
```
Could you remove this line:
`         let args = cli::cli()?;`

cli argument parsing is done in `src/main.rs`.

Davejkane commented on src/program/mod.rs:
```diff
@@ -31,6 +32,7 @@ pub struct Program<'a> {
     screen: Screen<'a>,
     paths: Paths,
     ui_state: ui::State,
+    sorter: Sorter,
```
Do we need to store the sorter in here? I think we'll only need to use it once, so there's probably no need to keep it for the life of the program.

NickHackman replied:
I thought the goal was to have it ready so that it would be usable in execution?
Though, I guess it's easy to just add it back in or somewhere else when using it at run time.

NickHackman replied:
Fixed

Davejkane commented on src/sort.rs:
```diff
@@ -0,0 +1,121 @@
+        if self.reverse {
+            pathes.reverse();
+        }
+        pathes
```
It this a deliberate spelling choice because `paths` conflicts with something?

NickHackman replied:
Yeah, that's a typo...

Davejkane reviewed (changes requested):
Just a couple of minor points from me.

NickHackman committed:
Fixed typo, removed Sorter from Program struct

NickHackman committed:
Merge branch 'development' into cli-sorting-options

Davejkane reviewed (approved):
Looks awesome. Thanks for the great contribution.

Davejkane commented:
@gurgalex can you please see if you're happy with this now?

NickHackman commented:
@Davejkane sorry, let me do one more push, just thought of a micro optimization

NickHackman committed:
Micro optimizations

gurgalex commented:
Looks good. Thanks again for the contribution.
